source/common/router/config_impl.cc

```diff
@@ -2147,13 +2146,42 @@ PerFilterConfigs::PerFilterConfigs(
     const Protobuf::Map<std::string, ProtobufWkt::Any>& typed_configs,
     const OptionalHttpFilters& optional_http_filters,
     Server::Configuration::ServerFactoryContext& factory_context,
-    ProtobufMessage::ValidationVisitor& validator) {
-  for (const auto& it : typed_configs) {
-    const auto& name = it.first;
-    auto object = createRouteSpecificFilterConfig(name, it.second, optional_http_filters,
-                                                  factory_context, validator);
-    if (object != nullptr) {
-      configs_[name] = std::move(object);
+    ProtobufMessage::ValidationVisitor& validator)
+    : ignore_optional_option_from_hcm_for_route_config_(Runtime::runtimeFeatureEnabled(
+          "envoy.reloadable_features.ignore_optional_option_from_hcm_for_route_config")) {
+
+  static const std::string filter_config_type =
+      envoy::config::route::v3::FilterConfig::default_instance().GetDescriptor()->full_name();
+
+  for (const auto& per_filter_config : typed_configs) {
+    const std::string& name = per_filter_config.first;
+    RouteSpecificFilterConfigConstSharedPtr config;
+
+    // Ignore the optional_http_filters if the ignore_optional_option_from_hcm_for_route_config_
+    // is set to true by the runtime feature.
+    bool is_optional_by_hcm = !ignore_optional_option_from_hcm_for_route_config_ &&
+                              (optional_http_filters.find(name) != optional_http_filters.end());
```

Comment:
Do you think you could elaborate on what's going on here a bit more? It's bit hard to follow the flow and the motivation for the different flows

Reply:
I will add more comment about this feature, thanks.